src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -172,20 +163,27 @@ private[concurrent] object BatchingExecutorStatics {
     }
   }
 
-  protected def unbatchedExecute(r: Runnable): Unit
-
-  private[this] final def batchedExecute(runnable: Runnable): Unit = {
-    val b = _tasksLocal.get
-    if (b ne null) b.push(runnable)
-    else unbatchedExecute(new Batch(runnable))
-  }
-
-  override def execute(runnable: Runnable): Unit =
-    if(batchable(runnable)) batchedExecute(runnable)
-    else unbatchedExecute(runnable)
-
-  /** Override this to define which runnables will be batched.
-    * By default it tests the Runnable for being an instance of [Batchable].
-   **/
+  protected def submitAsync(r: Runnable): Unit
+  protected def isAsync: Boolean = true
   protected def batchable(runnable: Runnable): Boolean = runnable.isInstanceOf[Batchable]
+
+  protected def reportFailure(throwable: Throwable): Unit
+
+  override final def execute(runnable: Runnable): Unit =
+    if (isAsync) {
+      if (batchable(runnable)) {
+        val b = _tasksLocal.get
+        if (b.isInstanceOf[Batch]) b.asInstanceOf[Batch].push(runnable)
+        else submitAsync(new Batch(runnable, resubmitOnBlock = true))
+      } else submitAsync(runnable)
+    } else {
+      Objects.requireNonNull(runnable) // Make sure we're not trying to add nulls
+      val b = _tasksLocal.get
+      if (b.isInstanceOf[Batch]) b.asInstanceOf[Batch].push(runnable)
+      else if (b ne null) {
+        _tasksLocal.set("")     // Set a marker to indicate that we are submitting synchronously
```

Comment:
change the `""` to a constant?

Reply:
`""` is already a constant baked into the bytecode

Reply:
yeah, the meaning might be clearer if it was a constant called `marker` or `sentinel`

Reply:
"" is a constant in the constant pool. I wanted something which is allocation-free and belongs to the root class loader in case it isn't cleared. (Since that is benign and does not leak memory)